File: lib/poro_entity_tuplizer.rb
class PoroEntityTuplizer
  include Java::OrgHibernateTupleEntity::EntityTuplizer

  # classes = [Java::OrgHibernateTupleEntity::EntityMetamodel, Java::OrgHibernateMapping::PersistentClass].to_java(Java::JavaLang::Class)
  # PoroEntityTuplizer.java_class.getDeclaredConstructor(classes)

  java_signature('PoroEntityTuplizer(org.hibernate.tuple.entity.EntityMetamodel entityMetamodel,
                                     org.hibernate.mapping.PersistentClass mappedEntity)')
  def initialize(entity_metamodel, mapped_entity)
  end

  java_signature 'java.lang.Object[] getPropertyValues(java.lang.Object entity)'
  def getPropertyValues(entity)
    fail NotImplementedError
  end

  java_signature 'void setPropertyValues(java.lang.Object entity, java.lang.Object[] values)'
  def setPropertyValues(entity, values)
    fail NotImplementedError
  end

  java_signature 'boolean isInstance(java.lang.Object object)'
  def isInstance(object)
    fail NotImplementedError
  end

  java_signature 'java.lang.Class getMappedClass()'
  def getMappedClass
    fail NotImplementedError
  end

  java_signature 'org.hibernate.property.Getter getGetter(int i)'
  def getGetter(index)
    fail NotImplementedError
  end

  java_signature 'org.hibernate.EntityMode getEntityMode()'
  def getEntityMode
    fail NotImplementedError
  end

  java_signature 'java.lang.Object instantiate()'
  java_signature 'java.lang.Object instantiate(java.io.Serializable id,
                                               org.hibernate.engine.spi.SessionImplementor session)'
  def instantiate(id, session = nil)
    fail NotImplementedError
  end

  java_signature 'java.io.Serializable getIdentifier(java.lang.Object entity,
                                                     org.hibernate.engine.spi.SessionImplementor session)'
  def getIdentifier(entity, session = nil)
    fail NotImplementedError
  end

  java_signature 'void setIdentifier(java.lang.Object entity,
                                     java.io.Serializable id,
                                     org.hibernate.engine.spi.SessionImplementor session)'
  def setIdentifier(entity, id, session = nil)
    fail NotImplementedError
  end

  java_signature 'void resetIdentifier(java.lang.Object entity,
                                       java.io.Serializable currentId,
                                       java.lang.Object currentVersion,
                                       org.hibernate.engine.spi.SessionImplementor session)'
  def resetIdentifier(entity, current_id, current_version, session = nil)
    fail NotImplementedError
  end

  java_signature 'java.lang.Object getVersion(java.lang.Object entity) throws org.hibernate.HibernateException'
  def getVersion(entity)
    fail NotImplementedError
  end

  java_signature 'void setPropertyValue(java.lang.Object entity, int i, java.lang.Object value)
                    throws org.hibernate.HibernateException'
  java_signature 'void setPropertyValue(java.lang.Object entity,
                                        java.lang.String propertyName,
                                        java.lang.Object value)
                    throws org.hibernate.HibernateException'
  def setPropertyValue(entity, property_name_or_index, value)
    fail NotImplementedError
  end

  java_signature 'java.lang.Object[] getPropertyValuesToInsert(java.lang.Object entity,
                                                               java.util.Map mergeMap,
                                                               org.hibernate.engine.spi.SessionImplementor session)'
  def getPropertyValuesToInsert(entity, merge_map, session)
    fail NotImplementedError
  end

  java_signature 'java.lang.Object getPropertyValue(java.lang.Object entity, int i)'
  java_signature 'java.lang.Object getPropertyValue(java.lang.Object entity, java.lang.String propertyName)
                    throws org.hibernate.HibernateException'
  def getPropertyValue(entity, property_name_or_index)
    fail NotImplementedError
  end

  java_signature 'void afterInitialize(java.lang.Object entity,
                                       boolean lazyPropertiesAreUnfetched,
                                       org.hibernate.engine.spi.SessionImplementor session)'
  def afterInitialize(entity, lazy_poperties_are_unfetched, session)
    fail NotImplementedError
  end

  java_signature 'boolean hasProxy()'
  def hasProxy
    false.to_java
  end

  java_signature 'java.lang.Object createProxy(java.io.Serializable id,
                                               org.hibernate.engine.spi.SessionImplementor session)
                    throws org.hibernate.HibernateException'
  def createProxy(id, session)
    fail NotImplementedError
  end

  java_signature 'boolean isLifecycleImplementor()'
  def isLifecycleImplementor
    false.to_java
  end

  java_signature 'java.lang.Class getConcreteProxyClass()'
  def getConcreteProxyClass
    fail NotImplementedError
  end

  java_signature 'boolean hasUninitializedLazyProperties(java.lang.Object entity)'
  def hasUninitializedLazyProperties(entity)
    fail NotImplementedError
  end

  java_signature 'boolean isInstrumented()'
  def isInstrumented
    false.to_java
  end

  java_signature 'org.hibernate.EntityNameResolver[] getEntityNameResolvers()'
  def getEntityNameResolvers
    fail NotImplementedError
  end

  java_signature 'java.lang.String determineConcreteSubclassEntityName(
                                     java.lang.Object entityInstance,
                                     org.hibernate.engine.spi.SessionFactoryImplementor factory)'
  def determineConcreteSubclassEntityName(entity_instance, factory)
    fail NotImplementedError
  end

  java_signature 'org.hibernate.property.Getter getIdentifierGetter()'
  def getIdentifierGetter
    fail NotImplementedError
  end

  java_signature 'org.hibernate.property.Getter getVersionGetter()'
  def getVersionGetter
    fail NotImplementedError
  end

  become_java!(false)
end

File: lib/java/src/org/hibernaterb/PoroEntityTuplizer.java
package org.hibernaterb;

import java.io.Serializable;
import java.lang.Override;
import java.util.Map;
import java.util.Set;
import java.util.HashSet;
import java.util.Iterator;

import org.hibernate.tuple.entity.AbstractEntityTuplizer;
import org.hibernate.mapping.PersistentClass;
import org.hibernate.metamodel.binding.EntityBinding;
import org.hibernate.metamodel.binding.AttributeBinding;
import org.hibernate.tuple.Instantiator;
import org.hibernate.tuple.entity.EntityMetamodel;
import org.hibernate.HibernateException;
import org.hibernate.mapping.Property;
import org.hibernate.property.Getter;
import org.hibernate.property.Setter;
import org.hibernate.proxy.ProxyFactory;
import org.hibernate.engine.spi.SessionFactoryImplementor;
import org.hibernate.EntityMode;
import org.hibernate.EntityNameResolver;
import org.hibernate.property.PropertyAccessor;
import org.hibernate.property.PropertyAccessorFactory;

import org.jruby.RubyObject;
import org.jruby.RubyClass;

import javax.script.ScriptContext;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;

import org.hibernaterb.PoroInstantiator;
import org.hibernaterb.PoroPropertyAccessor;

class PoroEntityTuplizer extends AbstractEntityTuplizer {
    private static ScriptEngine rubyEngine = new ScriptEngineManager().getEngineByName("jruby");

    private final Set lazyPropertyNames = new HashSet();

    public PoroEntityTuplizer(EntityMetamodel entityMetamodel, PersistentClass mappedEntity) {
        super(entityMetamodel, mappedEntity);

        Iterator iter = mappedEntity.getPropertyClosureIterator();
        while (iter.hasNext()) {
            Property property = (Property) iter.next();
            if (property.isLazy()) {
                lazyPropertyNames.add(property.getName());
            }
        }

        String[] getterNames = new String[propertySpan];
        String[] setterNames = new String[propertySpan];

        for (int i = 0; i < propertySpan; i++) {
            getterNames[i] = getters[i].getMethodName();
            setterNames[i] = setters[i].getMethodName();
        }
    }

    public PoroEntityTuplizer(EntityMetamodel entityMetamodel, EntityBinding mappedEntity) {
        super(entityMetamodel, mappedEntity);

        for (AttributeBinding property : mappedEntity.getAttributeBindingClosure()) {
            if (property.isLazy()) {
                lazyPropertyNames.add(property.getAttribute().getName());
            }
        }

        String[] getterNames = new String[propertySpan];
        String[] setterNames = new String[propertySpan];

        for (int i = 0; i < propertySpan; i++) {
            getterNames[i] = getters[i].getMethodName();
            setterNames[i] = setters[i].getMethodName();
        }
    }

    @Override
    public EntityMode getEntityMode() {
        return EntityMode.POJO;
    }

    @Override
    protected Getter buildPropertyGetter(Property mappedProperty, PersistentClass mappedEntity) {
        return mappedProperty.getPropertyAccessor(null).getGetter(null, mappedProperty.getName());
    }

    @Override
    protected Getter buildPropertyGetter(AttributeBinding mappedProperty) {
        return (new PoroPropertyAccessor()).getGetter(null, mappedProperty.getAttribute().getName());
    }

    @Override
    protected Setter buildPropertySetter(Property mappedProperty, PersistentClass mappedEntity) {
        return mappedProperty.getPropertyAccessor(null).getSetter(null, mappedProperty.getName());
    }

    @Override
    protected Setter buildPropertySetter(AttributeBinding mappedProperty) {
        return (new PoroPropertyAccessor()).getSetter(null, mappedProperty.getAttribute().getName());
    }

    @Override
    protected Instantiator buildInstantiator(PersistentClass mappingInfo) {
        return new PoroInstantiator(mappingInfo);
    }

    @Override
    protected Instantiator buildInstantiator(EntityBinding mappingInfo) {
        return new PoroInstantiator(mappingInfo);
    }

    @Override
    protected ProxyFactory buildProxyFactory(PersistentClass mappingInfo, Getter idGetter, Setter idSetter) {
        return null;
    }

    @Override
    protected ProxyFactory buildProxyFactory(EntityBinding mappingInfo, Getter idGetter, Setter idSetter) {
        return null;
    }

    @Override
    public Class getMappedClass() {
        return RubyObject.class;
    }

    @Override
    public Class getConcreteProxyClass() {
        return RubyObject.class;
    }

    @Override
    public boolean isInstrumented() {
        return false;
    }

    @Override
    public EntityNameResolver[] getEntityNameResolvers() {
        return new EntityNameResolver[] { BasicEntityNameResolver.INSTANCE };
    }

    @Override
    public String determineConcreteSubclassEntityName(Object entityInstance, SessionFactoryImplementor factory) {
        return extractEmbeddedEntityName((RubyObject) entityInstance);
    }

    public static String extractEmbeddedEntityName(RubyObject entity) {
        return entity.getMetaClass().getName();
    }

    public static class BasicEntityNameResolver implements EntityNameResolver {
        public static final BasicEntityNameResolver INSTANCE = new BasicEntityNameResolver();

        @Override
        public String resolveEntityName(Object entity) {
            if (!RubyObject.class.isInstance(entity)) return null;

            final String entityName = extractEmbeddedEntityName((RubyObject) entity );

            if (entityName == null) {
                throw new HibernateException( "Could not determine type of RubyObject entity" );
            }

            return entityName;
        }

        @Override
        public boolean equals(Object obj) {
            return getClass().equals( obj.getClass() );
        }

        @Override
        public int hashCode() {
            return getClass().hashCode();
        }
    }
}
